### PR TITLE
Fix to allow custom `CheckpointIO` with strategy classes

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Earlier since we initialized accelerator specific `CheckpointIO` in the strategy init, user's don't have an option to plugin custom `CheckpointIO` due to this:
https://github.com/Lightning-AI/lightning/blob/0e5312833fa70fb744f316f3db19c78b38eb9683/src/pytorch_lightning/trainer/connectors/accelerator_connector.py#L348-L352

Found while working on #13658

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @awaelchli @ananthsub @ninginthecloud @rohitgr7 @otaj @akihironitta

### CODE REVIEW

**Comment by rohitgr7** [comment]
need help with mypy

File: src/pytorch_lightning/strategies/strategy.py [hunk 1]
```diff
@@ -74,7 +74,10 @@ def accelerator(self, accelerator: "pl.accelerators.accelerator.Accelerator") ->
 
     @property
     def checkpoint_io(self) -> CheckpointIO:
-        return self._checkpoint_io if self._checkpoint_io is not None else TorchCheckpointIO()
+        if self._checkpoint_io is None:
+            self._checkpoint_io = TorchCheckpointIO()
```
Comment (otaj) [inline src/pytorch_lightning/strategies/strategy.py:78]:
```suggestion
            self._checkpoint_io: Optional[CheckpointIO] = TorchCheckpointIO()
```

This is from where mypy is getting the types that are throwing the errors. I fyou type this assignment, it works without errors. However, I think it might be nicer to have this variable typed already in `__init__`

Comment (rohitgr7) [reply src/pytorch_lightning/strategies/strategy.py:78]: it's typed in `__init__`, not sure why it's raising an error here

**Review by kaushikb11 — APPROVED** [review]

**Comment by kaushikb11** [comment]
Thanks!

File: src/pytorch_lightning/strategies/strategy.py (hunk 1 above)
Comment (otaj) [reply src/pytorch_lightning/strategies/strategy.py:78]: That is not true. What is typed in `__init__` is `self.checkpoint_io`, not `self._checkpoint_io`. That is then assigned through setter and I guess mypy does not follow through the setter

Comment (otaj) [reply src/pytorch_lightning/strategies/strategy.py:78]: the other option is to assign to `self._checkpoint_io` directly in `__init__` and not through the setter

Comment (rohitgr7) [reply src/pytorch_lightning/strategies/strategy.py:78]: ah! I thought it was `self._checkpoint_io: Optional[CheckpointIO]` all the time.

thank you :)

**Review by otaj — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/CHANGELOG.md [hunk 2]
```diff
@@ -345,6 +345,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `Trainer.predict(return_predictions=False)` to track prediction's batch_indices ([#13629](https://github.com/Lightning-AI/lightning/pull/13629))
 
 
+- Fixed and issue to allow custom `CheckpointIO` plugins with strategies ([#13785](https://github.com/Lightning-AI/lightning/pull/13785))
```
Comment (awaelchli) [inline src/pytorch_lightning/CHANGELOG.md:348]:
```suggestion
- Fixed and issue that prevented setting a custom `CheckpointIO` plugin with strategies ([#13785](https://github.com/Lightning-AI/lightning/pull/13785))
```

File: src/pytorch_lightning/strategies/hpu_parallel.py [hunk 3]
```diff
@@ -151,3 +151,13 @@ def teardown(self) -> None:
         # Was set to local rank
         os.environ.pop("ID", None)
         os.environ.pop("HCCL_DISTRIBUTED_BACKEND", None)
+
+    @property
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/hpu_parallel.py:155]: Please add the properties and setters to the top, after init :) 
Req. change for this so we don't diverge further from the current organization

File: tests/tests_pytorch/accelerators/test_accelerator_connector.py [hunk 4]
```diff
@@ -723,3 +724,12 @@ def test_plugin_only_one_instance_for_one_type(plugins, expected):
 def test_passing_zero_and_empty_list_to_devices_flag(accelerator, devices):
     with pytest.raises(MisconfigurationException, match="value is not a valid input using"):
         Trainer(accelerator=accelerator, devices=devices)
+
+
+@mock.patch("pytorch_lightning.strategies.hpu_parallel._HPU_AVAILABLE", return_value=False)
+@mock.patch("pytorch_lightning.accelerators.hpu._HPU_AVAILABLE", return_value=False)
+@mock.patch("pytorch_lightning.plugins.precision.hpu._HPU_AVAILABLE", return_value=False)
+def test_accelerator_specific_custom_plugin(*_):
```
Comment (awaelchli) [inline tests/tests_pytorch/accelerators/test_accelerator_connector.py:732]:
```suggestion
def test_accelerator_specific_checkpoint_io(*_):
```

**Review by awaelchli — CHANGES_REQUESTED** [review]

**Review by awaelchli — APPROVED** [review]
